File: examples/remote_artifact_quickstart.py
"""Verify a transport-neutral remote publication and persist its audit record."""

from __future__ import annotations

import argparse
import hashlib
from pathlib import Path

import rbfsafe


parser = argparse.ArgumentParser()
parser.add_argument("journal", type=Path, help="new transfer-journal directory")
args = parser.parse_args()

payload = b"immutable atlas payload\n"
item = rbfsafe.MemoryArtifactInput()
item.type = rbfsafe.MemoryArtifactType.SAFE_ATLAS
item.deployment_id = "arm-a"
item.robot_digest = "a" * 64
item.scene_digest = "b" * 64
item.task_id = "shelf-pick"
item.content_digest = hashlib.sha256(payload).hexdigest()
item.locator = "artifacts/shelf-atlas"
item.evidence = rbfsafe.EvidenceLevel.CERTIFIED_REGION
memory = rbfsafe.SafetyMemory()
artifact = memory.register_artifact(item)

# Demonstration key only. Production keys must come from an external secret manager.
key = bytes(range(1, 33))
request = rbfsafe.prepare_artifact_publish(
    memory,
    artifact.id,
    payload,
    "artifact-service",
    1,
    "application/vnd.rbfsafe.atlas",
)
receipt = rbfsafe.make_artifact_publish_receipt(request, 101)
receipt = rbfsafe.authenticate_artifact_publish_receipt(
    receipt, "service-key-1", key
)
verified = rbfsafe.verify_artifact_publish(
    memory, request, receipt, payload, "service-key-1", key
)
journal = rbfsafe.ArtifactTransferJournal()
record = journal.append(verified, "")
journal.save(args.journal)
loaded = rbfsafe.ArtifactTransferJournal.load(args.journal)

print(f"request={request.id}")
print(f"receipt={receipt.id}")
print(f"transfer={verified.id}")
print(f"record={record.id}")
print(f"journal={loaded.identity}")
print(
    "authentication="
    f"{rbfsafe.artifact_transfer_authentication_name(verified.authentication)}"
)
print("runtime_executable=false")

File: examples/artifacts/remote_artifact_quickstart.cpp
#include <rbfsafe/rbfsafe.h>

#include <array>
#include <filesystem>
#include <iostream>
#include <string>

namespace {

constexpr const char* kPayloadDigest = "4753e499617943c6b3dfa197752908e793797df2a669c7696ab3e53e534df4bd";

std::string digest(char value) { return std::string(64, value); }

rbfsafe::MemoryArtifactInput artifact_input() {
    rbfsafe::MemoryArtifactInput input;
    input.type = rbfsafe::MemoryArtifactType::SafeAtlas;
    input.deployment_id = "arm-a";
    input.robot_digest = digest('a');
    input.scene_digest = digest('b');
    input.task_id = "shelf-pick";
    input.content_digest = kPayloadDigest;
    input.locator = "artifacts/shelf-atlas";
    input.evidence = rbfsafe::EvidenceLevel::CertifiedRegion;
    return input;
}

std::array<std::byte, 32> example_key() {
    std::array<std::byte, 32> key{};
    for (std::size_t index = 0; index < key.size(); ++index)
        key[index] = static_cast<std::byte>(index + 1);
    return key;
}

} // namespace

int main(int argc, char** argv) {
    using namespace rbfsafe;
    if (argc != 2) {
        std::cerr << "usage: rbfsafe_remote_artifact_quickstart <new-journal-directory>\n";
        return 2;
    }
    const std::string payload_text = "immutable atlas payload\n";
    const auto payload = std::as_bytes(std::span(payload_text.data(), payload_text.size()));
    const auto key = example_key();

    SafetyMemory memory;
    auto artifact = memory.register_artifact(artifact_input());
    if (!artifact) {
        std::cerr << artifact.error().describe() << '\n';
        return 1;
    }
    auto request = prepare_artifact_publish(memory, artifact.value().id, payload, "artifact-service", 1,
                                            "application/vnd.rbfsafe.atlas");
    if (!request) {
        std::cerr << request.error().describe() << '\n';
        return 1;
    }
    auto unsigned_receipt = make_artifact_publish_receipt(request.value(), 101);
    if (!unsigned_receipt) {
        std::cerr << unsigned_receipt.error().describe() << '\n';
        return 1;
    }
    auto receipt = authenticate_artifact_publish_receipt(unsigned_receipt.value(), "service-key-1", key);
    if (!receipt) {
        std::cerr << receipt.error().describe() << '\n';
        return 1;
    }
    auto verified =
        verify_artifact_publish(memory, request.value(), receipt.value(), payload, "service-key-1", key);
    if (!verified) {
        std::cerr << verified.error().describe() << '\n';
        return 1;
    }

    ArtifactTransferJournal journal;
    auto record = journal.append(verified.value(), "");
    if (!record) {
        std::cerr << record.error().describe() << '\n';
        return 1;
    }
    auto saved = journal.save(std::filesystem::path(argv[1]));
    if (!saved) {
        std::cerr << saved.error().describe() << '\n';
        return 1;
    }
    auto loaded = ArtifactTransferJournal::load(std::filesystem::path(argv[1]));
    if (!loaded) {
        std::cerr << loaded.error().describe() << '\n';
        return 1;
    }
    std::cout << "request=" << request.value().id << '\n'
              << "receipt=" << receipt.value().id << '\n'
              << "transfer=" << verified.value().id << '\n'
              << "record=" << record.value().id << '\n'
              << "journal=" << loaded.value().identity() << '\n'
              << "authentication=" << artifact_transfer_authentication_name(verified.value().authentication)
              << '\n'
              << "runtime_executable=false\n";
    return 0;
}
